### PR TITLE
ARROW-13567: [C++] ConvertOptions::Defaults leaves `timestamp_parsers` uninitialized

### PR DESCRIPTION
**Pull request by westonpace**
Initialize `timestamp_parsers` to an empty list in the `ConvertOptions::Defaults` method

### CODE REVIEW

File: cpp/src/arrow/csv/options.cc [hunk 1]
```diff
@@ -43,6 +43,7 @@ ConvertOptions ConvertOptions::Defaults() {
                          "NULL", "NaN",  "n/a",      "nan",     "null"};
   options.true_values = {"1", "True", "TRUE", "true"};
   options.false_values = {"0", "False", "FALSE", "false"};
+  options.timestamp_parsers = {};
```
Comment (pitrou) [inline cpp/src/arrow/csv/options.cc:46]: Why is this necessary? `std::vector` is initialized to an empty vector by default.

**Comment by pitrou** [comment]
I'm afraid I don't understand this PR. @bkietz

**Comment by bkietz** [comment]
A default constructed vector will be initialized to empty. A vector with bogus or uninitialized size means that the underlying memory was corrupted, we're accessing a dangling reference to ConvertOptions, or similar undefined behavior. If you have solid repro I would recommend running it through valgrind or ASAN

**Comment by westonpace** [comment]
Ah, please forgive my foolishness.  I'll reopen if I find the root cause but now I'm wondering if I was just testing with R code built against an older version of Arrow.
